### PR TITLE
Newest updated

### PR DESCRIPTION
**Pull request by bruceadams**
This PR is working toward addressing #310. This is incomplete. There is a relevant test commented out because it fails.

### CODE REVIEW

**Comment by bruceadams** [comment]
This is pretty far from correct 😞 … some discussion on #310.

**Comment by bruceadams** [comment]
I'm much happier with this code. Pretty much everywhere that used to have some concept of max_version now has top_versions which contains both the highest semver version and the most recently updated version.

What is not yet here is _using_ the most recently updated version in the user interface. (I don't really know where this is in the code, but I haven't really looked either.)

File: src/models/krate.rs [hunk 1]
```diff
@@ -206,7 +214,7 @@ impl<'a> NewCrate<'a> {
 }
 
 impl Crate {
-    /// SQL filter based on whether the crate's name loosly matches the given
+    /// SQL filter based on whether the crate's name loosely matches the given
```
Comment (carols10cents) [inline src/models/krate.rs]: ❤️

File: src/tests/util.rs [hunk 2]
```diff
@@ -303,7 +303,7 @@ impl TestAppBuilder {
     }
 }
 
-/// A colleciton of helper methods for the 3 authentication types
+/// A collection of helper methods for the 3 authentication types
```
Comment (carols10cents) [inline src/tests/util.rs]: ❤️

File: src/controllers/krate/metadata.rs [hunk 3]
```diff
@@ -1,13 +1,10 @@
 //! Endpoints that expose metadata about a crate
 //!
-//! These endpoints provide data that could be obtained direclty from the
+//! These endpoints provide data that could be obtained directly from the
```
Comment (carols10cents) [inline src/controllers/krate/metadata.rs]: ❤️

File: src/models/version.rs [hunk 4]
```diff
@@ -39,6 +39,25 @@ pub struct NewVersion {
...
+/// The highest version (semver order) and the most recently updated version.
+/// Typically used for a single crate.
+#[derive(Debug, Clone)]
+pub struct TopVersions {
+    pub highest: semver::Version,
+    pub newest: semver::Version,
+}
+
+/// A default semver value, "0.0.0", for use in TopVersions
+fn default_semver_version() -> semver::Version {
```
Comment (carols10cents) [inline src/models/version.rs]: Nice extraction!

**Comment by jtgeibel** [comment]
This looks really good to me so far!  A few things I noticed running locally:

* The Just Updated column isn't showing a version, just empty `()`.  I think this may be causing the test failure as well.
* Everything still links to the top-level `crates/crate_name` route.  When a version is displayed, it would be nice if the link went directly to that version's page.

To test the end-to-end behavior, I'm looking to publish the following set of crates to staging (`Cargo.toml` below).  Let me know if you think of any additional edge cases that should be covered.

```toml
[workspace]

# The comments show the intended publish order
members = [
    "prerelease_only",        # 0.0.1-alpha.0
    "prerelease_then_normal", # 0.0.1-alpha.0, 0.0.1
    "normal_then_prerelease", # 0.0.1, 0.1.0-alpha.0
    "normal_release_only",    # 0.0.1
    "normal_releases",        # 0.0.1, 0.1.0, 0.1.1, 0.2.0
    "normal_and_backport",    # 0.1.0, 0.2.0, 0.2.1, 0.1.1
    "normal_and_backport2",   # 0.1.0, 0.2.0, 0.1.1, 0.2.1
]
```

**Comment by bruceadams** [comment]
I see the empty version for Just Updated and struggled to understand how `max_version` worked and `newest_version` did not. I think I have _that_ detail fixed at this point 🎉. I like suggestion about linking straight to a specific version when we display a specific version. I may not get to that today.

File: app/templates/index.hbs [hunk 5]
```diff
@@ -38,19 +38,19 @@
...
   </section>
   <section id='just-updated' data-test-just-updated aria-busy="{{dataTask.isRunning}}">
     <h2>Just Updated</h2>
-    {{crate-list crates=model.just_updated}}
+    {{crate-list-newest crates=model.just_updated}}
   </section>
   <section id='most-recently-downloaded' data-test-most-recently-downloaded aria-busy="{{dataTask.isRunning}}">
     <h2>Most Recent Downloads</h2>
-    {{crate-list crates=model.most_recently_downloaded}}
+    {{crate-list-max crates=model.most_recently_downloaded}}
```
Comment (carols10cents) [inline app/templates/index.hbs]: The "Most Recent Downloads" list is for any version of the crate, like "Most Downloaded" is, so I think this should use `crate-list-name-only` too? What do you think?

File: app/templates/components/crate-list-newest.hbs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<ol>
+  {{#each crates as |crate index|}}
+    <li>
+      {{#link-to 'crate.version' crate.id crate.newest_version class='name' data-test-crate-link=index}}
```
Comment (carols10cents) [inline app/templates/components/crate-list-newest.hbs]: So the "Just Updated" list shows a version and links to that version, and the "Most downloaded" lists don't show a version and don't link to a version... I wonder if, for consistency, we should make the "New Crates" list items link to the version because it displays a version? It probably doesn't matter in practice because new crates are only likely to have one version, but what do you think?

Comment (bruceadams) [reply app/templates/components/crate-list-newest.hbs]: I wondered about this. It felt weird to link to a specific version when we expect there to be only one. What do you think of not displaying the version number in the "New Crates" list? The meaning of _version number_ isn't very deep when there is only one.

File: app/templates/index.hbs (hunk 5 above)
Comment (bruceadams) [reply app/templates/index.hbs]: Ah! Right. I hadn't managed to see this. Thanks!

**Comment by bruceadams** [comment]
So… I'm happy with the state of the code and the user interface in this pull request as it currently stands. The main thing I'm uncomfortable about is that the scope of this change has grown well beyond the original issue #310. I am more than willing to trim this back to a narrow scope that only fiddles with the *Just Updated* column and leaves everything else as it is today. I can imagine some in between places as well. Preferences?

File: app/templates/components/crate-list-newest.hbs (hunk 6 above)
Comment (carols10cents) [reply app/templates/components/crate-list-newest.hbs]: Hmmm. Hmmm... the only slight pause I have is that the version number a new crate starts with is a small signal... if it's 0.0.1, it's probably someone's experiment... if it's 1.0.0, it's the launch of something someone's probably been working on for a while and wants to signal that people should use it... is that an interesting enough reason to keep the display of the version number for new crates? I don't know!

Comment (bruceadams) [reply app/templates/components/crate-list-newest.hbs]: Yes. I can see your point about version number being a signal for new crates. In addition, having the version number visible for new crates doesn't strike me as detrimental on the crates.io home page. So: my remaining question is just: what do we link to?
- Linking to the specific version aligns with the change in this PR makes for _Just Updates_, making link behavior more consistent.
- Linking to the front page of for the crate seems simpler; it gives a shorter url and a smidge less complexity. If someone sat looking at the crates.io front page for a while, then clicked on a new crate, there is a chance that crate had already published another version. Linking to the crate home would make that newer version more visible.

Comment (bruceadams) [reply app/templates/components/crate-list-newest.hbs]: I went with linking to the specific version under *New Crates*. That feels most right to me: consistent with *Just Updated* and I think my stated concern about yet newer versions isn't worth worrying about. The link the user clicks on takes them to what the text says, which, I think, meets the principle of least astonishment.

**Comment by bruceadams** [comment]
Sorry about the test failure 😞. Running the front end test locally stopped working and I haven't figured out why.

File: tests/acceptance/crate-test.js [hunk 7]
```diff
@@ -66,7 +66,7 @@ module('Acceptance | crate page', function(hooks) {
     await visit('/');
     await click('[data-test-just-updated] [data-test-crate-link="0"]');
 
-    assert.equal(currentURL(), '/crates/nanomsg');
+    assert.equal(currentURL(), '/crates/nanomsg/6.8.9');
```
Comment (carols10cents) [inline tests/acceptance/crate-test.js]: Is this test one that's failing for you locally? I think it's because this test uses the factories, and the version number isn't guaranteed to be the same every time the factory gets called?

This can be fixed by explicitly setting a `newest_version` for this particular instance, sort of like [this test explicitly sets a `max_version`](https://github.com/rust-lang/crates.io/blob/3e5dd28127284e64ff8e4b106ca5dc3472398e5a/tests/acceptance/crate-test.js#L29).

Comment (bruceadams) [reply tests/acceptance/crate-test.js]: Running the tests locally crashes fairly badly. I don't have much experience with npm stuffs, so I don't have much clue what is going on. It _did_ work for me. I don't know of having changed anything (although I'm a compulsive `brew upgrade`r and that might've broken something).

**Comment by bruceadams** [comment]
I will be at the Rust coffee this evening 🎉  @jtgeibel and I looked at the Percy failures. It seems to be related to auto generated stuff that has interesting cross cutting impact. In some bigger picture, it'd be great to stabilize that stuff.

I'm fine with test cleanup. I'm not going to be able to really look at code for several hours.

File: tests/acceptance/crate-test.js (hunk 7 above)
Comment (carols10cents) [reply tests/acceptance/crate-test.js]: Yikes! I have occasional luck with removing `node_modules` and re-`npm install`ing? We can look at it tonight though :)

Comment (jtgeibel) [reply tests/acceptance/crate-test.js]: Locally, I've replaced this test with the following.  This should eliminate the dependence on faker data and make the test more robust to future changes.  Would you add a final commit with this or a similar update?

```js
  test('visiting a crate page from the front page', async function(assert) {
    this.server.create('crate', { id: 'nanomsg', newest_version: '0.6.1' });
    this.server.create('version', { crate: 'nanomsg', num: '0.6.1' });

    await visit('/');
    await click('[data-test-just-updated] [data-test-crate-link="0"]');

    assert.equal(currentURL(), '/crates/nanomsg/0.6.1');
    assert.equal(title(), 'nanomsg - crates.io: Rust Package Registry');

    assert.dom('[data-test-heading] [data-test-crate-name]').hasText('nanomsg');
    assert.dom('[data-test-heading] [data-test-crate-version]').hasText('0.6.1');
  });
```

**Comment by bruceadams** [comment]
@jtgeibel Your test adjustment makes sense; I added it. The first rebase yesterday took me a while! I was happy the second one was easy!

**Comment by jtgeibel** [comment]
Thanks @bruceadams! I hate to say it, but this needs another rebase now, hopeful another easy one though.  If you get around to it today, ping me and I'll r+ ASAP.  There's been a lot of work on the frontend recently, sorry for the high rate of churn.

**Comment by bruceadams** [comment]
I learned more about Ember.js, including that the [documentation is very good](https://guides.emberjs.com/release/routing/linking-between-routes/#toc_the-linkto--component) even for a reader like me who barely has a clue about Ember.js.

**Comment by jtgeibel** [comment]
Thanks Bruce!

@bors: r+
